12-offer-encoding.md

```diff
@@ -0,0 +1,823 @@
+    2. data:
+        * [`tu64`:`min`]
+    1. type: 24 (`quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 30 (`node_id`)
+    2. data:
+        * [`point32`:`node_id`]
+    1. type: 54 (`send_invoice`)
+    1. type: 34 (`refund_for`)
+    2. data:
+        * [`sha256`:`refunded_payment_hash`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+1. subtype: `blinded_path`
+2. data:
+   * [`point`:`blinding`]
+   * [`byte`:`num_hops`]
+   * [`num_hops*onionmsg_path`:`path`]
+
+## Requirements For Offers
+
+A writer of an offer:
+  - if it sets `node_id`:
+    - MUST set `node_id` to the public key of the node to request the invoice from.
+  - otherwise:
+    - MUST provide at least one `blinded_path`
+    - MUST use the final `onionmsg_path` `point` in the first `blinded_path` as the implied `node_id` for `signature`.
+  - MAY specify exactly one signature TLV: `signature`:
+    - If so, it MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
+  - MUST set `description` to a complete description of the purpose
+    of the payment.
+  - if the chain for the invoice is not solely bitcoin:
+    - MUST specify `chains` the offer is valid for.
+  - otherwise:
+    - MAY omit `chains`, implying that bitcoin is only chain.
+  - if a specific minimum `amount` is required for successful payment:
+    - MUST set `amount` to the amount expected (per item).
+    - if the currency for `amount` is that of the first entry in `chains`:
+      - MUST specify `amount` in multiples of the minimum lightning-payable unit
+        (e.g. milli-satoshis for bitcoin).
+    - otherwise:
+      - MUST specify `iso4217` as an ISO 4712 three-letter code.
+      - MUST specify `amount` in the currency unit adjusted by the ISO 4712
+        exponent (e.g. USD cents).
+  - if it supports offer features:
+    - SHOULD set `features` to the bitmap of offer features.
+  - if the offer expires:
+    - MUST set `absolute_expiry` `seconds_from_epoch` to the number of seconds
+      after midnight 1 January 1970, UTC that invoice_request should not be
+      attempted.
+  - if it is connected only by private channels:
+    - MUST include `paths` containing one or more paths to the node from
+      publicly reachable nodes.
+  - otherwise:
+    - MAY include `paths`.
+  - if it includes `paths`:
+    - if the `node_id` is not public:
+      - MUST ignore any invoice_request which does not use one of the `paths`.
+    - otherwise:
+      - MAY ignore any invoice_request which does not use one of the `paths`.
+    - MUST NOT blind the first `onionmsg_path` `node_id` point.
+  - if it sets `issuer`:
+    - SHOULD set it to clearly identify the issuer of the invoice.
+    - if it includes a domain name:
+      - SHOULD begin it with either user@domain or domain
+      - MAY follow with a space and more text
+  - if it can supply more than one item for a single invoice
```

Comment:
> if it can supply more than one item for a single invoice

Does this mean "can supply more than one invoice for a single offer" or is it referring to physical stock of some item being sold?

Reply:
It means that you can buy several items with a single invoice.

If you create an offer for tips, a single-use offer or an offer for selling a unique item (a piece of art for instance), quantity is irrelevant and you should not set `quantity_min` or `quantity_max`.
If you sell items for which several copies exist and you're willing to sell multiple at once (gift cards or anything you can buy in a supermarket), then the buyer may want to buy several of them at once. In that case the offer must say how many of them you're willing to sell at once using `quantity_min` and `quantity_max`. Then the buyer will send an invoice request for the quantity that they want and the seller will return a single invoice for that quantity.

Reply:
My original reason for asking was thinking that this could possibly be simplifed to a single `quantity` field, which is interpreted as "<= quantity" (or 1, if not set). Specifically because I couldn't really see a case where `quantity_min` is not set to 1 - by this example it would be an offer that _only_ allows you to buy two gift cards at a time? In the case of a grocery store, I'd imagine that you have multiple different items, and then quantity doesn't apply? 

To me, this is a strange detail to include in the specification - it's simultaneously too much (covering a very specific enforced minimum buy with a range, rather than just a simple `quantity` field) and not enough (not covering a basket of distinct goods).

Reply:
You can't buy 1g of coffee beans?